paddle/lib/parameter_optimizer.h

```diff
@@ -0,0 +1,35 @@
+#ifndef PADDLE_LIB_OPTIMIZER_BASE_H_
+#define PADDLE_LIB_OPTIMIZER_BASE_H_
+
+#include <string>
+#include <functional>
+// #include <math/Tensor.h>
+#include "Tensor.h"
+
+namespace paddle {
+namespace lib {
+
+class ParameterOptimizer {
+public:
+  /*! \brief update hook for algorithm need to traverse parameter more
+    than once.
+  */
+  typedef std::function<void(Tensor&, const Tensor&)> UpdateHook;
+
+  static ParameterOptimizer *create();
+  virtual update(Tensor &parameter, const Tensor &gradient, double learning_rate) = 0;
```

Comment:
The C API uses `void*` buffer to pass in parameter and gradient, can Tensor be created from a raw buffer without a copy (not so familiar with C++ myself)